pkg/operator/kubeletversionskewcontroller/kubelet_version_skew_controller.go

```diff
@@ -0,0 +1,138 @@
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+const (
+	KubeletVersionSkewLimitUpgradeableConditionType = "KubeletVersionSkewLimitUpgradeable"
+
+	KubeletVersionUnknownReason           = "KubeletVersionUnknown"
+	KubeletVersionSyncedReason            = "KubeletVersionSynced"
+	KubeletVersionSkewLessThanLimitReason = "KubeletVersionSkewLessThanLimit"
+	KubeletVersionSkewAtLimitReason       = "KubeletVersionSkewAtLimit"
+	KubeletVersionUnsupportedReason       = "KubeletVersionUnsupported"
+)
+
+// KubeletVersionSkewController sets Upgradeable=False if the kubelet
+// version on a node falls too far behind the api server version.
+type KubeletVersionSkewController interface {
+	factory.Controller
+}
+
+func NewKubeletVersionSkewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	recorder events.Recorder,
+) *kubeletVersionSkewController {
+	c := &kubeletVersionSkewController{
+		operatorClient:   operatorClient,
+		nodeLister:       kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Lister(),
+		apiServerVersion: semver.MustParse(status.VersionForOperandFromEnv()),
+	}
+	// check if major version updated and no one updated this code
+	if c.apiServerVersion.Minor < 2 {
+		panic(fmt.Errorf("unable to determine kubelet version skew limit"))
+	}
+	c.Controller = factory.New().
+		WithSync(c.sync).
+		WithInformers(kubeInformersForNamespaces.InformersFor("").Core().V1().Nodes().Informer()).
+		ToController("KubeletVersionSkewController", recorder.WithComponentSuffix("kubelet-version-skew-controller"))
+	return c
+}
+
+type kubeletVersionSkewController struct {
+	factory.Controller
+	operatorClient   v1helpers.OperatorClient
+	nodeLister       corev1listers.NodeLister
+	apiServerVersion semver.Version
+}
+
+func (c *kubeletVersionSkewController) sync(_ context.Context, _ factory.SyncContext) error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(operatorSpec.ManagementState) {
+		return nil
+	}
+
+	nodes, err := c.nodeLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	var errorMsgs []string
+	var skewedToTheLimitMsgs []string
+	var skewedButOKMsgs []string
+	var syncedMsgs []string
+	var unsupportedMsgs []string
+
+	// for each node, check kubelet version
+	for _, node := range nodes {
+		kubeletVersion, err := nodeKubeletVersion(node)
+		skew := int(kubeletVersion.Minor - c.apiServerVersion.Minor)
```

Comment:
what if `kubeletVersion` is nil due to an `err`? If we always have a `kubeletVersion` then maybe we want `nodeKubeletVersion(node *corev1.Node) semver.Version`?

Reply:
Updated.